### PR TITLE
GitWorkingTree: Properly close the repository after each operation

### PR DESCRIPTION
**Pull request by sschuberth**
Previously, the `Repository` instance remained open during the lifetime
of the `GitWorkingTree`. That is a problem if some clean-up code (like
in tests) wants to delete a cloned Git repository while there is still
a GitWorkingTree instance.

To avoid that, provide a getter for the `repo` property which allows the
caller to auto-close the `Repository` instance after use as on the next
access of the `repo` property it will be re-evaluated. The added runtime
overhead due to this should be neglible compared to the duration of the
Git operation.

Signed-off-by: Sebastian Schuberth <sschuberth@gmail.com>

### CODE REVIEW

File: downloader/src/main/kotlin/vcs/GitWorkingTree.kt [hunk 1]
```diff
@@ -53,11 +53,12 @@ private fun findGitOrSubmoduleDir(workingDirOrFile: File): Repository {
 open class GitWorkingTree(workingDir: File, vcsType: VcsType) : WorkingTree(workingDir, vcsType) {
     companion object : Logging
 
-    private val repo = findGitOrSubmoduleDir(workingDir)
+    private val repo: Repository
```
Comment (fviernau) [inline downloader/src/main/kotlin/vcs/GitWorkingTree.kt:56]: IIUC the constructor threw an exception previously, and as of now not anymore.
Shall we ensure it fails early again?

File: downloader/src/main/kotlin/vcs/GitWorkingTree.kt [hunk 2]
```diff
@@ -53,11 +53,12 @@ private fun findGitOrSubmoduleDir(workingDirOrFile: File): Repository {
...
+        get() = findGitOrSubmoduleDir(workingDir)
 
-    override fun isValid(): Boolean = repo.objectDatabase?.exists() == true
+    override fun isValid(): Boolean = repo.use { it.objectDatabase?.exists() == true }
```
Comment (fviernau) [inline downloader/src/main/kotlin/vcs/GitWorkingTree.kt:59]: Commit-msg: The Repository class / interface seems to have an internal `useCount`, with the initial value of 1.
The class seems to release the resources when the count reaches 0. 

From the commit message I do not understand why the releasing of the resources is now triggered.
In particular what the actual trigger now is. Can you re-phrase it to explain that?
In particular, it should cover why the trigger is guaranteed to take place before the mentioned test clean-ups.

File: downloader/src/main/kotlin/vcs/GitWorkingTree.kt [hunk 3]
```diff
@@ -89,41 +90,55 @@ open class GitWorkingTree(workingDir: File, vcsType: VcsType) : WorkingTree(work
     }
 
     override fun getNested(): Map<String, VcsInfo> =
-        listSubmodulePaths(repo).associateWith { GitWorkingTree(repo.workTree.resolve(it), vcsType).getInfo() }
+        repo.use {
```
Comment (fviernau) [inline downloader/src/main/kotlin/vcs/GitWorkingTree.kt:93]: Have you considered exposing the same mechanism as the Repository class, e.g. `open` and `close()` functions instead? I believe with such approach it would be easier to move towards thread safety (hm, may we need that in the future? if not irrelevant). Note that not all subclasses of `Repository` are thread safe but the `FileRepository` is.

File: downloader/src/main/kotlin/vcs/GitWorkingTree.kt [hunk 4]
```diff
@@ -89,41 +90,55 @@ open class GitWorkingTree(workingDir: File, vcsType: VcsType) : WorkingTree(work
...
 
-    override fun getRootPath(): File = repo.workTree
+    override fun getRootPath(): File = repo.use { it.workTree }
 
     override fun listRemoteBranches(): List<String> =
-        runCatching {
-            LsRemoteCommand(repo).setHeads(true).call().map { it.name.removePrefix("refs/heads/") }
-        }.getOrElse {
-            throw IOException("Unable to list the remote branches.", it)
+        repo.use {
```
Comment (fviernau) [inline downloader/src/main/kotlin/vcs/GitWorkingTree.kt:124]: The diffs of this and below function contain a couple of unrelated changes.

**Review by fviernau — CHANGES_REQUESTED** [review]

File: downloader/src/main/kotlin/vcs/GitWorkingTree.kt [hunk 5]
```diff
@@ -53,11 +53,12 @@ private fun findGitOrSubmoduleDir(workingDirOrFile: File): Repository {
...
 
-    private val repo = findGitOrSubmoduleDir(workingDir)
+    private val repo: Repository
+        get() = findGitOrSubmoduleDir(workingDir)
 
-    override fun isValid(): Boolean = repo.objectDatabase?.exists() == true
+    override fun isValid(): Boolean = repo.use { it.objectDatabase?.exists() == true }
 
-    override fun isShallow(): Boolean = repo.directory?.resolve("shallow")?.isFile == true
+    override fun isShallow(): Boolean = repo.use { it.directory?.resolve("shallow")?.isFile == true }
```
Comment (fviernau) [inline downloader/src/main/kotlin/vcs/GitWorkingTree.kt:61]: Consider factoring out `repo.use` to a function.

File: downloader/src/main/kotlin/vcs/GitWorkingTree.kt (hunk 3 above)
Comment (sschuberth) [reply downloader/src/main/kotlin/vcs/GitWorkingTree.kt:93]: > Have you considered exposing the same mechanism as the Repository class, e.g. incrementOpen and close() functions instead?

No, I'd prefer to not reimplement that logic here.

> Note that not all subclasses of `Repository` are thread safe but the `FileRepository` is.

Aren't we good then already in terms of thread safety? Anyway, the goal of this PR is not thread-safety, but to avoid a resource leak. Let's do one topic at a time.

File: downloader/src/main/kotlin/vcs/GitWorkingTree.kt (hunk 5 above)
Comment (sschuberth) [reply downloader/src/main/kotlin/vcs/GitWorkingTree.kt:61]: Sorry, no. I'm not going to put an 8-letters-call to a function (whose name would probably longer).

File: downloader/src/main/kotlin/vcs/GitWorkingTree.kt (hunk 4 above)
Comment (sschuberth) [reply downloader/src/main/kotlin/vcs/GitWorkingTree.kt:124]: Unfortunately, you're not very specific which "unrelated" changes you refer to. So once again, I can only guess that you mean the renaming of lambda arguments.

This first renaming to `branch` becomes necessary in order to not shadow the `it` from `use`. The second renaming to `e` was indeed not complained about by the IDE, but IMO it should, as keeping the original code would also shadow the outer `it`.

File: downloader/src/main/kotlin/vcs/GitWorkingTree.kt (hunk 1 above)
Comment (sschuberth) [reply downloader/src/main/kotlin/vcs/GitWorkingTree.kt:56]: I actually think it's even better the new way: There might be cases where you'd like to create a `GitWorkingTree` instance that points to a `workingDir` that yet needs to be initialized (maybe even in the scope of file existence checks in rules).

File: downloader/src/main/kotlin/vcs/GitWorkingTree.kt (hunk 4 above)
Comment (fviernau) [reply downloader/src/main/kotlin/vcs/GitWorkingTree.kt:124]: I didn't notice the parameter shadowing, so I'm fine with the changes.

File: downloader/src/main/kotlin/vcs/GitWorkingTree.kt (hunk 3 above)
Comment (fviernau) [reply downloader/src/main/kotlin/vcs/GitWorkingTree.kt:93]: > Anyway, the goal of this PR is not thread-safety, but to avoid a resource leak. Let's do one topic at a time.

I was a bit unclear. My goal was not to ask you to put thread safety in scope of the PR. My intention was rather to bring it to attention and to clarify if we think it will be needed in future. Anyhow, I'm convinced now that it won't make moving towards thread safety (much) more difficult...

File: downloader/src/main/kotlin/vcs/GitWorkingTree.kt (hunk 1 above)
Comment (fviernau) [reply downloader/src/main/kotlin/vcs/GitWorkingTree.kt:56]: Hm, I believe failing early is good by default unless there is a particular case which satisfies not doing so.
I'd prefer keeping fail early until we come across a case which requires it's removal. But am not going to block this.

**Review by fviernau — APPROVED** [review]

File: downloader/src/main/kotlin/vcs/GitWorkingTree.kt (hunk 1 above)
Comment (sschuberth) [reply downloader/src/main/kotlin/vcs/GitWorkingTree.kt:56]: > I believe failing early is good by default

I agree, but what I'm trying to say is that creating a `GitWorkingTree` instance wrapping a directory that is no Git tree yet is not necessarily a failure. It only has to be a Git tree once a Git operation is executed on it.

That's BTW also how the other (purely CLI-based) `WorkingTree` implementations work: They will only throw an exception at the time a CLI tool is run and fails, not at instance construction time. So this commits also aligns the behavior across `WorkingTree` implementations.
